Source/Managers/DFU/FirmwareUpgradeManager.swift

```diff
@@ -520,39 +566,49 @@ public class FirmwareUpgradeManager : FirmwareUpgradeController, ConnectionObser
             return
         }
         // Check that the image array exists.
-        guard let images = response.images, images.count > 0 else {
+        guard let responseImages = response.images, responseImages.count > 0 else {
             self.fail(error: FirmwareUpgradeError.invalidResponse(response))
             return
         }
         
         switch self.mode {
         case .confirmOnly:
             // The new image should be in slot 1.
-            if images.count != 2 {
+            if responseImages.count != self.images.count {
                 self.fail(error: FirmwareUpgradeError.invalidResponse(response))
                 return
             }
             // Check that the new image is in permanent state.
-            if !images[1].permanent {
-                self.fail(error: FirmwareUpgradeError.unknown("Image is not in a permanent state."))
+            if !responseImages[self.i + 1].permanent {
+                self.fail(error: FirmwareUpgradeError.unknown("Image \(self.i + 1) is not in a permanent state."))
                 return
             }
-            // Image was confirmed, reset the device.
-            self.reset()
+            if self.i == self.images.count - 1 {
+                // Image was confirmed, reset the device.
+                self.reset()
+            } else {
+                self.i += 1
+                self.upload()
+            }
         case .testAndConfirm:
             // Check that the upgrade image has successfully booted.
-            if Data(images[0].hash) != self.hash {
+            if Data(responseImages[self.i].hash) != self.images[self.i].hash {
                 self.fail(error: FirmwareUpgradeError.unknown("Device failed to boot into new image."))
                 return
             }
             // Check that the new image is in confirmed state.
-            if !images[0].confirmed {
+            if !responseImages[self.i].confirmed {
                 self.fail(error: FirmwareUpgradeError.unknown("Image is not in a confirmed state."))
                 return
             }
-            // Confirm successful.
-            self.log(msg: "Upgrade complete", atLevel: .application)
-            self.success()
+            if self.i == self.images.count - 1 {
+                // Confirm successful.
+                self.log(msg: "Upgrade complete", atLevel: .application)
+                self.success()
+            } else {
+                self.i += 1
```

Comment:
Does it mean, that you confirm first image and than upload second image?
The process should be like that:
1. upload all images
2. send _test_ (if `mode` is `.testOnly` or `testAndConfirm`) or _confirm_ (if `.confirmOnly`) messages for all images
3. reset
4. send _confirm_ message is not sent before

Reply:
I wish I knew myself